internal/localtypes/file_permission_type.go

```diff
@@ -0,0 +1,67 @@
+package localtypes
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr/xattr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+var (
+	_ basetypes.StringTypable = FilePermissionType{}
+	_ xattr.TypeWithValidate  = FilePermissionType{}
+)
+
+type FilePermissionType struct {
+	basetypes.StringType
+}
+
+func NewFilePermissionType() FilePermissionType {
+	return FilePermissionType{StringType: types.StringType}
+}
+
+// Validate checks that the given input string is a valid file permission string,
+// expressed in numeric notation.
+// See: https://en.wikipedia.org/wiki/File-system_permissions#Numeric_notation
+func (f FilePermissionType) Validate(ctx context.Context, value tftypes.Value, path path.Path) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if value.IsNull() {
+		return diags
+	}
+
+	if !value.IsKnown() {
+		return diags
+	}
+
+	var fp string
+	err := value.As(&fp)
```

Comment:
Hopefully we can nix this `tftypes`-based handling via https://github.com/hashicorp/terraform-plugin-framework/issues/589 in the future.